serializer/streaming_json.rst

```diff
@@ -0,0 +1,672 @@
+        use Symfony\Component\TypeInfo\Type;
+
+        class TombolaService
+        {
+            private StreamReaderInterface $jsonStreamReader;
+            private string $catsJsonFile;
+
+            public function __construct(
+                private string $catsJsonFile,
+            ) {
+                $this->jsonStreamReader = JsonStreamReader::create();
+            }
+
+            public function pickTheTenthCat(): ?Cat
+            {
+                $jsonResource = fopen($this->catsJsonFile, 'r');
+                $type = Type::iterable(Type::object(Cat::class));
+
+                /** @var iterable<Cat> $cats */
+                $cats = $this->jsonStreamReader->read($jsonResource, $type);
+
+                $i = 0;
+                foreach ($cats as $cat) {
+                    if ($i === 9) {
+                        return $cat;
+                    }
+
+                    ++$i;
+                }
+
+                return null;
+            }
+
+            /**
+             * @return list<string>
+             */
+            public function listEligibleCatNames(): array
+            {
+                $jsonResource = fopen($this->catsJsonFile, 'r');
+                $type = Type::iterable(Type::object(Cat::class));
+
+                /** @var iterable<Cat> $cats */
+                $cats = $this->jsonStreamReader->read($jsonResource, $type);
+
+                return array_map(fn(Cat $cat) => $cat->name, iterator_to_array($cats));
+            }
+        }
+
+.. note::
+
+    You can explicitly inject the ``json_streamer.stream_reader`` service by
+    using the ``#[Target('json_streamer.stream_reader')]`` autowire attribute.
+
+The upper code demonstrates two different approaches to decoding JSON data
+using the JsonStreamer:
+
+* decoding from a stream (``pickTheTenthCat``)
+* decoding from a string (``listEligibleCatNames``).
+
+Both methods work with the same JSON data but differ in memory usage and
+speed optimization.
+
+
+Decoding from a stream
+~~~~~~~~~~~~~~~~~~~~~~
+
+In the ``pickTheTenthCat`` method, the JSON data is read as a stream using
+:phpfunction:`fopen`. Streams are useful when working with large files
+because the data is processed incrementally rather than loading the entire
+file into memory.
+
+To improve memory efficiency, the JsonStreamer creates `ghost objects`_
+instead of fully instantiating objects. Ghosts objects are lightweight
+placeholders that represent the objects but don't fully load their data
+into memory until it's needed. This approach reduces memory usage, especially
+for large datasets.
+
+* Advantage: Efficient memory usage, suitable for very large JSON files.
+* Disadvantage: Slightly slower than decoding a full string because data is
+  loaded on-demand.
+
+Decoding from a string
+~~~~~~~~~~~~~~~~~~~~~~
+
+In the ``listEligibleCatNames`` method, the entire JSON file is read into
+a string using :phpfunction:`file_get_contents`. This string is then passed
```

Comment:
I don't see `file_get_contents` usage but `fopen`. Is it outdated ?

Reply:
Indeed, the code needed to be updated, nice catch :slightly_smiling_face: